### PR TITLE
Netsim: Put back instructions dialog

### PR DESCRIPTION
**Pull request by islemaster**
Fixes a regression introduced in https://github.com/code-dot-org/code-dot-org/pull/11533: We stopped displaying an instructions dialog when opening NetSim levels as part of the cleanup around the topInstructions transition, but since NetSim does not support topInstructions we need to get that dialog back.  A related regression (not showing the instructions dialog when clicking on the small instructions) probably dates from further back.

This is a quick fix for those two missing behaviors, that has NetSim now understanding that it's a one-off when it comes to showing the instructions dialog (the widget levels are like this too, so it's not the only case).  We show the instructions entirely through changes to NetSim code, not touching StudioApp.  This whole 'level front-matter' system is badly in need of an overhaul, but this PR isn't the right place to do it - so forgive some clunky workarounds here.

I've also added a scenario to the NetSim UI test to specifically check for the two restored behaviors (now that they are a NetSim-specific concern) and done some general cleanup of test steps along the way.  I'll make notes on those inline.

### CODE REVIEW

File: dashboard/test/ui/step_definitions/netsim.rb [hunk 1]
```diff
@@ -1,17 +1,16 @@
...
+# Given I am on the 1st NetSim test level
+# Navigates to the requested level in the AllTheThings script.
+Given /^I am on the (\d+)(?:st|nd|rd|th)? (?:Net ?Sim|Internet Simulator) test level$/ do |level_index|
   steps <<-STEPS
-    And I am on "http://studio.code.org/s/allthethings/stage/14/puzzle/1?noautoplay=true"
-    And I wait to see "#netsim-lobby-name"
+    And I am on "http://studio.code.org/s/allthethings/stage/#{NETSIM_ALLTHETHINGS_STAGE}/puzzle/#{level_index}?noautoplay=true"
+    And I rotate to landscape
   STEPS
-  steps "And I close the dialog" if @browser.find_element(:id => 'x-close') rescue false
```
Comment (islemaster) [inline dashboard/test/ui/step_definitions/netsim.rb:14]: This "close the dialog, rescue false" step is why the NetSim UI tests failed to catch this problem in the first place.  I'm not sure what I was thinking at this point - maybe I was having a hard time reliably closing the dialog, or sometimes there wasn't one?

File: dashboard/test/ui/step_definitions/steps.rb [hunk 2]
```diff
@@ -131,9 +131,20 @@ def install_js_error_recorder
   wait_with_timeout.until { @browser.execute_script("return $(#{selector.dump}).text();").include? text }
 end
 
-When /^I wait until element "([^"]*)" is visible$/ do |selector|
+def jquery_is_element_visible(selector)
+  "return $(#{selector.dump}).is(':visible') && $(#{selector.dump}).css('visibility') !== 'hidden';"
```
Comment (islemaster) [inline dashboard/test/ui/step_definitions/steps.rb:135]: We have an inconsistent definition of "visible" spread around our UI tests; we still do, somewhat, but I clean it up quite a bit here.  In many places we were only checking `$(selector).is(':visible')`, but in a few others we also checked `$(selector).css('visibility') !== 'hidden'`; I guess the one isn't covered by the other.  This led me to the odd discovery that I could "Wait until dialog is visible" then assert "dialog is visible" and fail on the second step :sob:.

The changes I've made prefer checking both where we used to check one, and tests are passing with that change.

File: dashboard/test/ui/step_definitions/steps.rb [hunk 3]
```diff
@@ -538,20 +545,28 @@ def type_into_selector(input_text, selector)
...
+  @browser.execute_script(jquery_element_exists(selector))
+end
+
+def element_visible?(selector)
+  @browser.execute_script(jquery_is_element_visible(selector))
+end
+
 Then /^element "([^"]*)" is (not )?visible$/ do |selector, negation|
-  visibility = @browser.execute_script("return $(#{selector.dump}).css('visibility')")
-  visible = @browser.execute_script("return $(#{selector.dump}).is(':visible')") && (visibility != 'hidden')
```
Comment (islemaster) [inline dashboard/test/ui/step_definitions/steps.rb:543]: Here we checked both conditions, but for some reason were making two round trips to the remote browser to do so.  This has been reduced to a single expression and a single request to the remote browser in the refactor.  Every little bit counts.

**Comment by islemaster** [comment]
I'm actually not sure how beneficial that will be; there are a lot of possible initial views for NetSim depending on whether you're signed in, how many sections you belong to, etc so we'd need several eyes tests to cover all of them.  In addition most of the intialPageViews tests close the instructions dialog before taking a screenshot, so I think adding NetSim to that test would not have caught this particular regression.

Independent of the instructions dialog, do you feel like we need NetSim eyes tests?  I'm tempted to say its unit and UI test coverage is pretty good and it may not be worth the additional maintenance.

**Comment by Hamms** [comment]
Ahh, very good point. Not to mention that NetSim's comprehensive unit test coverage is significantly faster than an additional eyes test would be; I agree that additional eyes tests would likely be more trouble than they're worth.

**Pull request merged**